File: kotlin/custom-set/src/main/kotlin/CustomSet.kt
class CustomSet(vararg values: Int) {

    val items = values.toMutableSet()

    fun isEmpty(): Boolean = items.isEmpty()

    fun isSubset(other: CustomSet): Boolean = items.all { item -> other.contains(item) }

    fun isDisjoint(other: CustomSet): Boolean = items.none { item -> other.contains(item) }

    fun contains(other: Int): Boolean = items.contains(other)

    fun intersection(other: CustomSet): CustomSet = CustomSet(*items.filter { it in other.items }.toIntArray())

    fun add(other: Int) {
        items.add(other)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is CustomSet) return false

        return other.items == this.items
    }

    operator fun plus(other: CustomSet): CustomSet = CustomSet(*(items + other.items).toIntArray())

    operator fun minus(other: CustomSet): CustomSet = CustomSet(*(items - other.items).toIntArray())
}

File: csharp/custom-set/CustomSet.cs
public class CustomSet(params int[] values)
{
    private readonly HashSet<int> values = [.. values];

    public CustomSet Add(int value)
    {
        values.Add(value);
        return this;
    }

    public bool Empty() => values.Count == 0;

    public bool Contains(int value) => values.Contains(value);

    public bool Subset(CustomSet right) => values.All(right.Contains);

    public bool Disjoint(CustomSet right) => !values.Any(right.Contains);

    public CustomSet Intersection(CustomSet right) => new([.. values.Intersect(right.values)]);

    public CustomSet Difference(CustomSet right) => new([.. values.Where(v => !right.Contains(v))]);

    public CustomSet Union(CustomSet right) => new([.. values.Union(right.values)]);

    public override bool Equals(object? obj) => obj is CustomSet other && Subset(other) && other.Subset(this);

    public override int GetHashCode() => HashCode.Combine(values);
}